File: src/main/java/dmgcalculator/publisher/CardHoverPublisher.java
package dmgcalculator.publisher;

import com.megacrit.cardcrawl.cards.AbstractCard;

import java.util.ArrayList;
import java.util.List;

import dmgcalculator.interfaces.OnCardHoveredSubscriber;

public class CardHoverPublisher {

    private static final List<OnCardHoveredSubscriber> onCardHoveredSubscribers = new ArrayList<>();

    private static AbstractCard hoveringCard;

    public static void publishOnCardHovered(AbstractCard card) {
        if (hoveringCard != card) {
            hoveringCard = card;
            onCardHoveredSubscribers.forEach(subscriber -> {
                subscriber.onCardHovered(card);
            });
        }
    }

    public static void publishNoCardHovering() {
        if (hoveringCard != null) {
            hoveringCard = null;
            onCardHoveredSubscribers.forEach(OnCardHoveredSubscriber::onNoCardHovering);
        }
    }

    public static void subscribe(OnCardHoveredSubscriber subscriber) {
        onCardHoveredSubscribers.add(subscriber);
    }
}

File: src/main/kotlin/dmgcalculator/publisher/CardHoverPublisher.kt
package dmgcalculator.publisher

import com.megacrit.cardcrawl.cards.AbstractCard
import dmgcalculator.interfaces.CardHoverSubscriber

object CardHoverPublisher {

    private val cardHoverSubscribers = mutableListOf<CardHoverSubscriber>()

    private var hoveredCard: AbstractCard? = null

    fun publishCardHovered(card: AbstractCard) {
        if (hoveredCard !== card) {
            hoveredCard = card
            cardHoverSubscribers.forEach {
                it.onCardHovered(card)
            }
        }
    }

    fun publishCardUnHovered() {
        if (hoveredCard != null) {
            hoveredCard = null
            cardHoverSubscribers.forEach {
                it.onCardUnHovered()
            }
        }
    }

    fun subscribe(subscriber: CardHoverSubscriber) {
        cardHoverSubscribers.add(subscriber)
    }
}
